framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,601 @@
+            }
+            printWriter.print(returnType);
+            printWriter.print(" ");
+        }
+        printWriter.print(methodName);
+        printWriter.print("(");
+
+        StringJoiner parameters = new StringJoiner(", ");
+
+        if (!aMethod.receiver.type.tlAnnotationsHere.isEmpty()
+                || !aMethod.receiver.type.innerTypes.isEmpty()) {
+            // Only output the receiver if it has an annotation.
+            parameters.add(formatAField(aMethod.receiver, "this", fullyQualifiedClassname));
+        }
+        for (Integer index : aMethod.parameters.keySet()) {
+            AField param = aMethod.parameters.get(index);
+            // AMethod doesn't actually track real parameter names.
+            // Fortunately, the stub file parser also doesn't, so
+            // this code can safely ignore that problem and use a generic
+            // name instead.
+            //
+            // TODO: use the actual parse tree of the method
+            // to figure out the real parameter names, and then thread them
+            // through to here.
+            parameters.add(formatAField(param, "param" + index, fullyQualifiedClassname));
+        }
+        printWriter.print(parameters.toString());
+        printWriter.println(");");
+    }
+
+    /**
+     * The implementation of {@link #write(AScene, Map, Map, Map, Writer)} and {@link #write(AScene,
+     * Map, Map, Map, String)}. Prints imports, classes, method signatures, and fields in stub file
+     * format, all with appropriate annotations.
+     *
+     * @param scene the scene to write
+     * @param printWriter where to write the scene to
+     */
+    private void writeImpl(AScene scene, PrintWriter printWriter) {
+        // Write out all imports
+        ImportDefCollector importDefCollector;
+        try {
+            importDefCollector = new ImportDefCollector(scene, printWriter);
+        } catch (DefException e) {
+            throw new BugInCF(e);
+        }
+        importDefCollector.visit();
+        printWriter.println();
+
+        // For each class
+        for (Map.Entry<String, AClass> classEntry : scene.classes.entrySet()) {
+            @SuppressWarnings("signature") // TODO: annotate AScene library
+            @BinaryName String classname = classEntry.getKey();
+            AClass aClass = classEntry.getValue();
+            String pkg = packagePart(classname);
+            String basename = basenamePart(classname);
+
+            // Do not attempt to print stubs for anonymous inner classes, because the stub parser
+            // cannot read them. (An anonymous inner class has a basename like Outer.1, so this
+            // check ensures that the binary name's final segment after its last . is not only
+            // composed of digits.)
+            if (basename.contains(".")) {
+                String innermostClassname = basename.substring(basename.lastIndexOf('.') + 1);
+                if (innermostClassname.matches("\\d+")) {
```

Comment:
Always precompute Patterns rather than representing them as strings, for efficiency.
Also, you can just use a regular expression here rather than using both `contains` and a regex.
Finally, you might as well do this computation on classname, before setting the local variables, for efficiency.